### ISSUE TITLE
Push Parser

### ISSUE THREAD

**Issue by kennethuil** [nikomatsakis/lalrpop#267]
If you're interested, I'm working on an implementation.

1. An enum ParserApi goes into the Algorithm struct.  It defaults to Pull, but will be set to Push if the grammar annotation "#[push]" is present.
2. For all implementations, a Parser struct is introduced to hold state.  Its impl gets two public functions, "feed" and "finish".  The parser implementation methods also become methods of that struct and take a &mut self.  A flag is added to the Parser struct to indicate whether the current chunk is the last chunk.
 * The introduction of a Parser struct can happen unconditionally; my tests indicate that LLVM optimizes out unused references to self, and there are plans for Rust to implement this optimization as well.  However, I'd have to measure if it has any noticeable impact on compile times.
3. For both implementations, parser code will need to treat the end of the input collection differently depending on whether "feed" or "finish" is being called.
4. For the parse table implementation, the local variables in the main parser loop method become fields of the Parser struct, and hold state between invocations of feed/finish.
 * If there is no error recovery, the shift loop can simply exit when it reaches the end of the chunk, and the next invocation of "feed" or "finish" will immediately reenter the loop and pick up where it left off.
 * If there is error recovery, the Parser struct will need a flag to indicate that error recovery is in progress.
5. For the recursive ascent implementation, the Parser struct will need fields to hold state in between chunks, and the state methods will need code inserted to handle saving and restoring state.

**Comment by ahmedcharles** [comment]
Is this a feature request? I'm trying to figure out how to make this more actionable.

**Comment by nikomatsakis** [comment]
@kennethuil I do think supporting a push mode would be great. However, I might be inclined to hold off on these changes and instead pursue https://github.com/lalrpop/lalrpop/issues/303 -- if we had that work done, creating a push parser would I think just be a matter of adding the relevant code to `lalrpop_util`.

**Comment by nikomatsakis** [comment]
> For the recursive ascent implementation, the Parser struct will need fields to hold state in between chunks, and the state methods will need code inserted to handle saving and restoring state.

I don't think it makes sense to support this for recursive ascent; or if we did so, we'd want to generate a future or something. The whole point there is to re-use the program stack, essentially, to maintain the state.
